File: compilado/src/pilhaArray.c
#include "../includes/libs.h"
#include "../includes/pilhaArray.h"

//push
void pushPilhaArray(PilhaArray* pilha, int n) {
    if(pilha->header < MAX_SIZE) { 
       pilha->arr[pilha->header++] = n; //adiciona n no topo da pilha caso ela nao esteja cheia
    }
    else {
        printf("pilha cheia! %d nao foi inserido..", n);
    }
}

//pop
int popPilhaArray(PilhaArray* pilha) {
    if (pilha->header > 0) {
        pilha->header--;
        return pilha->arr[pilha->header]; //retorna o valor atual no topo da pilha
    } else {
        return -1; //retorna -1 indicando que a pilha esta vazia
    }
}

//isEmpty
bool isEmptyPilhaArray(PilhaArray* pilha) {
    return (pilha->header == -1); //verifica de o header esta com o valor -1, indicando que a pilha esta vazia
    }

//len
int lenPilhaArray(PilhaArray* pilha) {
    return (pilha->header + 1); //retorna o tamanho da pilha
}

File: interpretado/pilhaArray.py
class PilhaArray:
    def __init__(self, capacity):  #construtor
        self.capacity = capacity
        self.arr = [None] * capacity
        self.header = -1
    
    def isEmpty(self):  
        return self.header == -1  #retorna true caso o header == 1, i.e., caso nenhum item esteja empilhado
    
    def isFull(self):
        return (self.header == self.capacity - 1)  #retorna true caso o header esteja prestes a estourar a capacidade do array
    
    def push(self, valor):  
        if self.isFull():
            print("erro! pilha cheia...")
            return
        self.header += 1
        self.arr[self.header] = valor  #define o proximo item da pilha como o valor inserido

    def pop(self):
        if self.isEmpty():
            print("erro! nenhum item na pilha..")
            return
        valor = self.arr[self.header]  #aloca o valor do topo da pilha em uma variavel local
        self.header -= 1  #diminui o header de maneira que ele "aponte" para o valor anterior da fila, que agora estara no topo
        return valor
    
    def len(self):
        return (self.header + 1)
